### ISSUE TITLE
RxMoya.TargetType vs Moya.TargetType

### ISSUE THREAD

**Issue by AndrewSB** [Moya/Moya#551]
I'm refactoring my networking code, and I've found a compiler error that I had previously (badly) worked around.

I have an `OnlineProvider` similar to the one in Artsy's iOS app. It's class definition looks like: 
`class OnlineProvider<Target where Target: RxMoya.TargetType>: RxMoyaProvider<Target>`

The part I don't like here, is the `Target where Target: RxMoya.TargetType` because it makes me define my `TargetType` enum as `enum MyAPI: RxMoya.TargetType`. I'd much rather have `Target where Target: Moya.TargetType` and `enum MyAPI: Moya.TargetType`

So in attempt to change that, I've been able to refactor `MyAPI` to conform to `Moya.TargetType`, but I'm now having an issue with my `OnlineProvider`:

<img width="975" alt="screen shot 2016-07-15 at 11 24 19 am" src="https://cloud.githubusercontent.com/assets/3814772/16884514/b69e0810-4a7e-11e6-8819-73343257f43f.png">
I'm not sure, but I *think* that's because `RxMoyaProvider` wants to be specialized with a `RxMoya.TargetType`, not a `Moya.TargetType` and because `RxMoya.TargetType` doesn't inherit from `Moya.TargetType`, the compiler treats them as totally separate protocols (with the exact same definition, albeit)?

**Comment by ashfurrow** [comment]
Yeah, seems likely. We've run into problems in the past on our Demo app (`import`ing more than one version of Moya). You're using Carthage, eh?

**Comment by AndrewSB** [comment]
unfortunately

assuming that if you're using cocoapods, there is no `RxMoya.TargetType` since cocoapods doesn't actually build modules?

**Comment by ashfurrow** [comment]
Well, CocoaPods will include all the specified subspecs in one module, named `Moya`. So if you have `pod 'Moya/RxSwift`, that will pull all the RxSwift Moya files _and_ the regular Moya files into one module. Carthage doesn't have an equivalent of subspecs (they're stance is that you should use multiple frameworks instead – which we considered but decided against).

_Maybe_ you could do something like

```swift
typealias TargetType = RxMoya.TargetType
```

And then you wouldn't see `RxMoya.`?

**Comment by AndrewSB** [comment]
@ashfurrow playing with this right now, will hopefully update soon with what I ended up doing

**Comment by AndrewSB** [comment]
I think the underlying issue (speaking from the perspective of the compiled code that Carthage gives us) is that you can't pass a Moya.TargetType to an RxMoyaProvider.

**Comment by petester42** [comment]
I had started doing something like that [here](https://github.com/Moya/Moya/tree/petester42/dependent-carthage-frameworks) which works great in Xcode but I was getting Carthage errors.

**Comment by AndrewSB** [comment]
@petester42 can you talk more about the Carthage errors? I feel as if this problem will be exasperated once people start using Moya through the swift package manager, so finding an answer before we're up against a wall would be 🔥

**Comment by petester42** [comment]
I don't really remember what they were but the build using Carthage would fail but building with Xcode worked fine.

**Comment by AndrewSB** [comment]
hey @petester42, saw that you merged `master` into your branch. Did you mean to rebase? It's still saying that your branch is ~120 commits behind master

I'll try to take a look at this right now

**Comment by petester42** [comment]
Its fine. I'm just playing around. Don't worry about it

**Comment by AndrewSB** [comment]
Discovery: 
When I try to manually run what carthage uses to build the RxMoya iOS scheme (which is `xcrun xcodebuild -project /Users/asb/Desktop/Bruh/Carthage/Checkouts/Moya/Moya.xcodeproj -scheme "RxMoya iOS" -configuration Release -sdk iphoneos ONLY_ACTIVE_ARCH=NO BITCODE_GENERATION_MODE=bitcode CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY=  CARTHAGE=YES clean build`) here's the output I get:

```
xcrun xcodebuild -project /Users/asb/Desktop/Bruh/Carthage/Checkouts/Moya/Moya.xcodeproj -scheme "RxMoya iOS" -configuration Release -sdk iphoneos ONLY_ACTIVE_ARCH=NO BITCODE_GENERATION_MODE=bitcode CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY=                        
Build settings from command line:
    BITCODE_GENERATION_MODE = bitcode
    CODE_SIGN_IDENTITY = 
    CODE_SIGNING_REQUIRED = NO
    ONLY_ACTIVE_ARCH = NO
    SDKROOT = iphoneos9.3

=== BUILD TARGET Moya tvOS OF PROJECT Moya WITH CONFIGURATION Release ===
...
...
...

CompileSwift fail
```

Where it says it's trying to build the `Moya tvOS` target, not the `RxMoya iOS` target that was passed into `xcrun` as an argument... 

Not sure why that's happening, but the other targets seem to work correctly - just `RxMoya iOS`. I'm going to check out what happens if I try the same `xcrun` on `master`.

**Comment by ashfurrow** [comment]
That's odd. Hmm. @Moya/contributors Anyone else seeing build errors with Carthage?

**Comment by AndrewSB** [comment]
Stranger still: If I do a clean clone of Moya/Moya `master`, and I try the same `xcrun` with the `RxMoya iOS` target, it works, major steps:
`=== CLEAN TARGET Moya iOS OF PROJECT Moya WITH CONFIGURATION Release ===`, 
`=== CLEAN TARGET RxMoya iOS OF PROJECT Moya WITH CONFIGURATION Release ===`,
`=== BUILD TARGET Moya iOS OF PROJECT Moya WITH CONFIGURATION Release ===`, and then
`=== BUILD TARGET RxMoya iOS OF PROJECT Moya WITH CONFIGURATION Release ===`.

But if I try @petester42's branch, it sometimes first cleans `RxMoya watchOS`, or `RxMoya tvOS`, and then fails while trying to build the same. Not sure why it's doing that.

I'm going to branch off master and see if my branch has the same behavior or if one of @petester42's commits introduced the inconsistent clean+build procedure

**Comment by AndrewSB** [comment]
@petester42 can you explain how https://github.com/Moya/Moya/commit/e1b02527c15b3920c9968511303f68358040d735 works?

What script toggles the imports on and off?

**Comment by petester42** [comment]
There is a prebuild and a postbuild script that is located in the build phase section of the targets. It is there because cocoapods doesn't need the extra import. By default the import should be commented out so cocoapods works and when building Carthage it will toggle it on, build, and then toggle it off again.

**Comment by AndrewSB** [comment]
gotcha, can you also touch upon the changes you made to the `project.pbxproj`? Reorganizing the files/ folder structure?

**Comment by petester42** [comment]
The changes are mainly removing the Moya files from the reactive targets.

**Comment by AndrewSB** [comment]
@petester42 you said "cocoapods doesn't need the extra import", does having the import cause a compile error for cocoa pods? or will it just be a redundant line of code? I feel as if it would be ok to live with the redundant line of code for cocoapod users to avoid to complexity of the build script

**Comment by petester42** [comment]
It won't build with cocoapods in you have `import Moya` anywhere because it is building Moya itself.

**Comment by AndrewSB** [comment]
that makes sense - ok. Script it is

**Comment by petester42** [comment]
I think this has been addressed by #563

**Comment by AndrewSB** [comment]
Yup! I did delete the pre & post build scripts in that PR 👍
